### PR TITLE
Replace D-Bus with HTTP in the WebSocket server

### PR DESCRIPTION
**Pull request by kraih**
This PR removes all D-Bus related code from the WebSocket server (`OpenQA::WebSockets`) and replaces it with plain old HTTP. It also introduces a new client library (`OpenQA::WebSockets::Client`), so we can have RPC code that was previously split up among model, controller and utility modules in a single place.

Progress: https://progress.opensuse.org/issues/46802

### CODE REVIEW

**Comment by kraih** [comment]
You can start reviewing the PR, but it will have to stay a WIP until i had time to test it on the staging machines.

**Comment by kraih** [comment]
Deploying to the staging machines has shown some interesting problems. I'll have to revisit localhost authentication, since `OpenQA::UserAgent` is reused by `OpenQA::WebSockets::Client`.

**Comment by kraih** [comment]
I did some more testing on staging-1 yesterday, and everything looks good now.

File: lib/OpenQA/WebSockets.pm [hunk 1]
```diff
@@ -14,82 +14,126 @@
...
+    my $self = shift;
+
+    $self->defaults(appname => 'openQA Websocket Server');
+    $self->mode('production');
+
+    # no cookies for worker, no secrets to protect
+    $self->secrets(['nosecretshere']);
+    $self->config->{no_localhost_auth} ||= 1;
+
+    # Assetpack is required to render layouts pages
```
Comment (Martchus) [inline lib/OpenQA/WebSockets.pm]: I'm aware that you've only moved this line but I'm wondering why the web socket server would render layout pages.

Comment (kraih) [reply lib/OpenQA/WebSockets.pm]: Yea, i've been trying very hard to avoid any logic changes during the refactoring. But AssetPack is indeed a pretty heavy dependency that seems useless here. I'll test removing it.

**Comment by kraih** [comment]
I'll resolve the merge conflicts later, since i want to add a few new tests first (for which you gave me the idea 😉).

With `one of the services if offline` i assume you mean the WebSockets server, since all interactions are initiated by other services to the WebSockets server with `OpenQA::WebSockets::Client`. And to answer that question, you will see errors raised like in [this Travis test run](https://travis-ci.org/kraih/openQA/jobs/517199543) (which i had to fix afterwards).
```
[ERROR] Failed dispatching message to websocket server over ipc for worker "host:": Connection refused at lib/OpenQA/WebSockets/Client.pm line 58.
```
With the new client library errors should be a lot more consistent, the old D-Bus code was really bad about this. I'm sure this can still be improved, but by default i made the new API raise meaningful exceptions whenever possible.

**Comment by Martchus** [comment]
> With one of the services is offline i assume you mean the WebSockets server, since all interactions are initiated by other services to the WebSockets server with OpenQA::WebSockets::Client.

Yes. And I think the error message is good enough. (I just wanted to be sure we don't see some strange Perl warnings like it was when introduction the worker cache.)

**Comment by kraih** [comment]
This PR should be ready now.

File: lib/OpenQA/Parser/Result/OpenQA.pm [hunk 2]
```diff
@@ -37,7 +37,7 @@ sub parsed_details {
 sub search_in_details {
     my ($self, $field, $re) = @_;
     my $results = OpenQA::Parser::Result::OpenQA::Results->new();
-    $results->add($_) for grep { $_->{$field} =~ $re } @{$self->details};
+    $results->add($_) for grep { ($_->{$field} // '') =~ $re } @{$self->details};
```
Comment (kraih) [inline lib/OpenQA/Parser/Result/OpenQA.pm]: This change is unrelated to the main topic of the PR. But i got annoyed with all the warnings it caused on Travis.

**Comment by Martchus** [comment]
I'm giving it also a try under my local Tumbleweed development setup.

**Comment by kraih** [comment]
That error is raised by the [result call here](https://github.com/os-autoinst/openQA/pull/2043/files#diff-725ec16907bbc0c0ad1062d59de695d1R61). If we have to we can wrap it in an eval in the future and amend the error message.

File: lib/OpenQA/Schema/ResultSet/DeveloperSessions.pm [hunk 3]
```diff
@@ -56,7 +56,8 @@ sub register {
     # inform the worker that a new the developer session has been started
     if ($result && !$is_session_already_existing) {
         # hope this IPC call isn't blocking too long (since the livehandler isn't preforking)
-        OpenQA::IPC->ipc->websockets('ws_send', $worker_id, 'developer_session_start', $job_id);
+        my $client = OpenQA::WebSockets::Client->singleton;
+        $client->send_msg($worker_id, 'developer_session_start', $job_id);
```
Comment (kalikiana) [inline lib/OpenQA/Schema/ResultSet/DeveloperSessions.pm]: I like how the code actually looks more readable now

File: lib/OpenQA/Schema/Result/Workers.pm [hunk 4]
```diff
@@ -236,27 +237,28 @@ sub info {
...
 sub send_command {
     my ($self, %args) = @_;
-    return if (!defined $args{command});
+    return undef if (!defined $args{command});
 
     if (!grep { $args{command} eq $_ } COMMANDS) {
         my $msg = 'Trying to issue unknown command "%s" for worker "%s:%n"';
         log_error(sprintf($msg, $args{command}, $self->host, $self->instance));
-        return;
+        return undef;
```
Comment (kalikiana) [inline lib/OpenQA/Schema/Result/Workers.pm]: Is there a functional difference? Just wondering... it definitely looks clearer.

Comment (jberger) [reply lib/OpenQA/Schema/Result/Workers.pm]: Explicit return undef avoids context sensitivity
